File: python/perfect_squares.py
# no. 279
class Solution:
    def numSquares(self, n: int) -> int:
        if n <= 3:
            return n
        squares = []
        dp = [0 for _ in range(n + 1)]
        for i in range(1, n):
            if i ** 2 > n:
                break
            else:
                squares.append(i ** 2)
                dp[i ** 2] = 1 

        for i in range(1, n+1):
            if not dp[i]:
                min_ = dp[i - 1]
                for num in squares:
                    if num > i:
                        break
                    min_ = min(min_, dp[i - num])
                dp[i] = min_ + 1
        return dp[-1]

File: C++/perfect_squares.cpp
// no. 279
#include <vector>
using namespace std;

class Solution {
public:
  int numSquares(int n) {
    if (n <= 3) {
      return n;
    }
    vector<int> squares;
    vector<int> dp;
    for (int i = 0; i <= n; i++) {
      dp.push_back(0);
    }
    for (int i = 1; i < n; i++) {
      if (i * i > n) {
        break;
      } else {
        squares.push_back(i * i);
        dp[i * i] = 1;
      }
    }
    for (int i = 1; i <= n; i++) {
      if (dp[i] == 0) {
        int min_ = dp[i - 1];
        for (int j : squares) {
          if (j > i) {
            break;
          }
          min_ = min(min_, dp[i - j]);
        }
        dp[i] = min_ + 1;
      }
    }
    return dp.back();
  }
};
